src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+            IAsyncCompletionSession session,
+            AsyncCompletionData.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+            => Task.FromResult(UpdateCompletionList(session, data, cancellationToken));
+
+        private AsyncCompletionData.FilteredCompletionModel UpdateCompletionList(
+            IAsyncCompletionSession session,
+            AsyncCompletionData.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            AssertIsBackground();
+
+            SubscribeEvents(session);
+            if (!session.Properties.TryGetProperty(CompletionSource.HasSuggestionItemOptions, out bool hasSuggestedItemOptions))
+            {
+                // This is the scenario when the session is created out of Roslyn, in some other provider, e.g. in Debugger.
+                // For now, the default hasSuggestedItemOptions is false. We can discuss if the opposite is required.
+                hasSuggestedItemOptions = false;
+            }
+
+            var filterText = session.ApplicableToSpan.GetText(data.Snapshot);
+            var reason = data.Trigger.Reason;
+
+            // We do not care about the character in the case. We care about the reason only.
+            var roslynTrigger = Helpers.GetRoslynTrigger(data.Trigger, data.Trigger.Character);
+
+            if (!session.Properties.TryGetProperty(CompletionSource.InitialTriggerKind, out CompletionTriggerKind initialRoslynTriggerKind))
+            {
+                initialRoslynTriggerKind = CompletionTriggerKind.Invoke;
+            }
+
+            // Check if the user is typing a number. If so, only proceed if it's a number
+            // directly after a <dot>. That's because it is actually reasonable for completion
+            // to be brought up after a <dot> and for the user to want to filter completion
+            // items based on a number that exists in the name of the item. However, when
+            // we are not after a dot (i.e. we're being brought up after <space> is typed)
+            // then we don't want to filter things. Consider the user writing:
+            //
+            //      dim i =<space>
+            //
+            // We'll bring up the completion list here (as VB has completion on <space>).
+            // If the user then types '3', we don't want to match against Int32.
+            if (filterText.Length > 0 && char.IsNumber(filterText[0]))
+            {
+                if (!IsAfterDot(data.Snapshot, session.ApplicableToSpan))
+                {
+                    session.Dismiss();
+                    return null;
+                }
+            }
+
+            // We need to filter if a non-empty strict subset of filters are selected
+            var selectedFilters = data.SelectedFilters.Where(f => f.IsSelected).Select(f => f.Filter).ToImmutableArray();
```

Comment:
is this the same semantics as: ComputeEffectiveFilterItemState?

Reply:
yes, it is the same semantic as ComputeEffectiveFilterItemState

---
In reply to: [232509742](https://github.com/dotnet/roslyn/pull/29016#discussion_r232509742) [](ancestors = 232509742)